src/Orleans.Pipeline.Client/OrleansPipe.cs

```diff
@@ -66,49 +71,122 @@ public async Task Stop(CancellationToken token)
         catch (Exception ex)
         {
-            logger.LogError(ex, "An error occurred during Stopping");
+            logger.LogError(ex, "An error occurred during stopping.");
         }
         finally
         {
-            _writerStoppingTokenSource.Dispose();
             _observerStoppingTokenSource.Dispose();
+            UpdateStatus(OrleansPipeStatus.Broken);
         }
     }
 
-    #endregion
+    public async Task<bool> TryWriteAsync(TToServer item, CancellationToken cancellationToken)
+    {
+        if (_status == OrleansPipeStatus.Broken)
+        {
+            throw new BrokenOrleansPipeException("Cannot write to a broken pipe.");
+        }
+
+        if ((DateTime.UtcNow.Ticks - Interlocked.Read(ref _lastHeartbeatTicks)) < ExpectedHeartbeatIntervalTicks)
+        {
+            await _writer.Writer.WriteAsync(item, cancellationToken);
+            return true;
+        }
+
+        if (!await _reconnectionSemaphore.WaitAsync(0, cancellationToken))
+        {
+            return false; // Another reconnection attempt is in progress.
+        }
+
+        try
+        {
+            UpdateStatus(OrleansPipeStatus.Recovering);
+
+            for (int i = 0; i < MaxReconnectionAttempts; i++)
+            {
+                try 
+                {
+                    await _grain.Subscribe(_observer);
+                }
+                catch 
+                {
+                    await Task.Delay(ReconnectionDelay, cancellationToken);
+                    continue;
+                }
+
+                Interlocked.Exchange(ref _lastHeartbeatTicks, DateTime.UtcNow.Ticks);
+                await _writer.Writer.WriteAsync(item, cancellationToken);
+
+                UpdateStatus(OrleansPipeStatus.Healthy);
+
+                return true;
+            }
+
+            await Stop(CancellationToken.None); // This will stop the reader too, so clients can gracefully give up on this pipe.
+            UpdateStatus(OrleansPipeStatus.Broken);
+
+            return false;
+        }
+        finally
+        {
+            _reconnectionSemaphore.Release();
+        }
+    }
+
+    public IAsyncEnumerable<TFromServer> ReadAllAsync(CancellationToken cancellationToken) =>
+        _reader.Reader.ReadAllAsync(cancellationToken);
 
-    #region IOrleansPipeObserver<TFromServer>
     public async Task ReceiveMessage(PipeTransferItem<TFromServer> message)
     {
-        if (message is { Mode: TransferMode.Data, Item: not null })
+        switch (message.Mode)
         {
-            await _reader.Writer.WriteAsync(message.Item);
+            case TransferMode.Data:
+                {
+                    ArgumentNullException.ThrowIfNull(message.Item, "Item must no be null using data transfer mode");
+                    await _reader.Writer.WriteAsync(message.Item);
+                }
+                break;
+            case TransferMode.Heartbeat:
+                Interlocked.Exchange(ref _lastHeartbeatTicks, DateTime.UtcNow.Ticks);
```

Comment:
The result for heartbeat is only evaluated on a write attempt. in mostly listen scenarios this would mean a long undetected broken pipe. i will create a issue to keep this in mind, but dont let this be a blocker for merge.